pkg/handler3/handler_test.go

```diff
@@ -120,3 +122,153 @@ func TestRegisterOpenAPIVersionedService(t *testing.T) {
+	buffer := new(bytes.Buffer)
+	if err := json.Compact(buffer, returnedOpenAPI); err != nil {
+		t.Errorf("%v", err)
+	}
+	compactOpenAPI := buffer.Bytes()
+	var hash = computeETag(compactOpenAPI)
+
+	json.Unmarshal(compactOpenAPI, &s)
+
+	returnedJSON, err := json.Marshal(s)
+	if err != nil {
+		t.Fatalf("Unexpected error in preparing returnedJSON: %v", err)
+	}
+
+	returnedPb, err := ToV3ProtoBinary(compactOpenAPI)
+
+	if err != nil {
+		t.Fatalf("Unexpected error in preparing returnedPb: %v", err)
+	}
+
+	mux := http.NewServeMux()
+	o, err := NewOpenAPIService(nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	mux.Handle("/openapi/v3", http.HandlerFunc(o.HandleDiscovery))
+	mux.Handle("/openapi/v3/apis/apps/v1", http.HandlerFunc(o.HandleGroupVersion))
+
+	o.UpdateGroupVersion("apis/apps/v1", s)
+
+	server := httptest.NewServer(mux)
+	defer server.Close()
+	client := server.Client()
+
+	tcs := []struct {
+		acceptHeader string
+		respStatus   int
+		urlPath      string
+		respBody     []byte
+		expectedHash string
+		cacheControl string
+	}{
+		// Correct hash should yield the proper expiry and Cache Control headers
+		{"application/json",
+			200,
+			"openapi/v3/apis/apps/v1?hash=" + hash,
+			returnedJSON,
+			hash,
+			"public, immutable",
+		},
+		{"application/com.github.proto-openapi.spec.v3@v1.0+protobuf",
+			200,
+			"openapi/v3/apis/apps/v1?hash=" + hash,
+			returnedPb,
+			hash,
+			"public, immutable",
+		},
+		// Incorrect hash should redirect to the page with the correct hash
+		{"application/json",
+			200,
+			"openapi/v3/apis/apps/v1?hash=" + "OUTDATEDHASH",
+			returnedJSON,
+			hash,
+			"public, immutable",
+		},
+		{"application/com.github.proto-openapi.spec.v3@v1.0+protobuf",
+			200,
+			"openapi/v3/apis/apps/v1?hash=" + "OUTDATEDHASH",
+			returnedPb,
+			hash,
+			"public, immutable",
+		},
+
+		// No hash should not return Cache Control information
+		{"application/json",
+			200,
+			"openapi/v3/apis/apps/v1",
+			returnedJSON,
+			"",
+			"",
+		},
+		{"application/com.github.proto-openapi.spec.v3@v1.0+protobuf",
+			200,
+			"openapi/v3/apis/apps/v1",
+			returnedPb,
+			"",
+			"",
+		},
+	}
+
+	for _, tc := range tcs {
+		req, err := http.NewRequest("GET", server.URL+"/"+tc.urlPath, nil)
+		if err != nil {
+			t.Errorf("Accept: %v: Unexpected error in creating new request: %v", tc.acceptHeader, err)
+		}
+
+		req.Header.Add("Accept", tc.acceptHeader)
+		resp, err := client.Do(req)
+		if err != nil {
+			t.Errorf("Accept: %v: Unexpected error in serving HTTP request: %v", tc.acceptHeader, err)
+		}
+
+		if resp.StatusCode != 200 {
+			t.Errorf("Accept: Unexpected response status code, want: %v, got: %v", 200, resp.StatusCode)
+		}
+
+		if cacheControl := resp.Header.Get("Cache-Control"); cacheControl != tc.cacheControl {
+			t.Errorf("Expected Cache Control %v, got %v", tc.cacheControl, cacheControl)
+		}
+
+		if tc.expectedHash != "" {
+			if hash := resp.Request.URL.Query()["hash"]; hash[0] != tc.expectedHash {
```

Comment:
What is this actually testing?

Reply:
I see, so this is testing to see if the redirect worked well?

Reply:
Checks if the hash parameter returned in the response matches the latest hash. If the client supplied the correct hash, this hash should remain the same. If the client supplied the incorrect hash, this should return the hash of the latest document.